### ISSUE TITLE
Serious performance issues in Sap Hana driver during insert

### ISSUE THREAD

**Issue by mathiasmaerker** [typeorm/typeorm#6520]
**Issue type:**

[ ] question
[x ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[x] sap hana
**TypeORM version:**

[x ] `latest`

**Steps to reproduce or a small repository showing the problem:**

While we developed some importing mechanism we discovered that importing into a hana database takes for ages. inserting 1K entries takes about 53 seconds. Doesn't really matter if we use `save()`or ìnsert()` and we talk about a very simple entity without any relations 
```javascript
@Entity({ name: 'p4c_labels' })
export class Label {
  @PrimaryColumn()
  locale: string
  @PrimaryColumn()
  labelId: string
  @Column()
  labelValue: string
  @Column()
  entityId: string
  @Column()
  active: boolean
}
```
in comparison, I did some test with the native `@sap/hana-client` which is underlying in typeorm as well and I inserted 40K entries in 8 seconds. I know that orm adds some overhead and I am totally fine with that but that much difference can't be normal

@Szbuli can you maybe provide any insight here? Since you wrote the majority of the driver? 

Best regards 
Mathias

**Comment by Szbuli** [comment]
At my company we had a custom sap driver, but the official driver was not written by me. I have only made some changes, ie. using the pool. You can try to increase the pool size. 
I can have a look at your code and see if that helps. Can you post the typeorm and the native driver code?

**Comment by Szbuli** [comment]
I am pretty sure that save will fire a select for the ids to check if the entity exists or not.
Try https://typeorm.io/#/insert-query-builder

**Comment by mathiasmaerker** [comment]
Hi @Szbuli ,
unfortunately the repo is private but I can ask if I can invite you. The pool options never worked for me but I can give you the code for the different drivers:

So the Entity is in my previous post.
I created a other issue because I think bulk Insert does not work, because Hana does not support it, you can check that in #6519 

Here is some dummy testing coding for **typeorm**
```javascript
      const cnt = query.createQueryBuilder().insert().into(Country, ['id', 'alpha2', 'alpha3'])
      console.log(`Time elapsed execution: ${(endTime - startTime) / 1000}`)
      startTime = Date.now()
      for (let index = 0; index < 500; index++) {
        await cnt.values({ id: index, alpha2: 'b2', alpha3: 'bl3' }).execute()
      }
      endTime = Date.now()
      console.log(`Time elapsed Coun try insert: ${endTime - startTime}`)
```
there I insert into a country table, it is a very simple table
I have to loop because of #6519 
takes for ages(50sec), `save()` is not better

```javascript

@Entity({ name: 'p4c_country' })
export class Country {
  @PrimaryColumn()
  id: number

  @Column({ length: 2 })
  alpha2: string
  @Column({ length: 3 })
  alpha3: string
}
```
In terms of connection options I have not much:
```javascript
  const options: SapConnectionOptions = {
    name: 'sap',
    type: 'sap',
    host: credentials.host,
    port: parseInt(credentials.port),
    username: loginConfig.username,
    password: loginConfig.password,
    logging: 'all',
    synchronize,
    encrypt: true,
    extra: {
      sslValidateCertificate: true,
      sslTrustStore: credentials.certificate,
      sslCryptoProvider: 'openssl',
    },
    schema: loginConfig.schema.toUpperCase(),
    entities: [entities],
  }
```
maybe you can provide a pool example?


**Native version:**
```javascript
    const conn = hana.createConnection()

    const conn_params = {
      host: '**dbaas.ondemand.com',
      port: 30650,
      uID: 'system',
      pwd: '***',
      currentSchema: 'P4C_JMM',
      encrypt: true,
      sslValidateCertificate: true,
      sslCryptoProvider: 'openssl',
      sslTrustStore:
        '-----BEGIN CERTIFICATE----....
    }

    conn.connect(conn_params, function (err) {
      if (err) throw err
      const startTime = Date.now()
      const stmt = conn.prepare('INSERT INTO "p4c_country" VALUES(?, ?, ?)')
      const arr = []
      for (let index = 2; index <= 5002; index++) {
        const a = [index, 'b2', 'bl3']
        arr.push(a)
      }
      const result = stmt.execBatch(arr)
      stmt.drop()
      const endTime = Date.now()
      console.log(`Elapsed Time: ${(endTime - startTime) / 1000}`)
      conn.disconnect()
...
```
this takes only some miliseconds to complete and here we have 5K instead of 500 entries.


Does it help?

thanks a lot for your help!

Regards
Mathias

**Comment by imnotjames** [comment]
Isn't the native one doing all the queries at once?  And the typeorm example awaits for every query to complete serially.

What if you collect all the promises without awaiting & then `Promise.all` them?

```
      const cnt = query.createQueryBuilder().insert().into(Country, ['id', 'alpha2', 'alpha3'])
      console.log(`Time elapsed execution: ${(endTime - startTime) / 1000}`)
      startTime = Date.now()
      const queryPromises = []
      for (let index = 0; index < 500; index++) {
        queryPromises.push( cnt.values({ id: index, alpha2: 'b2', alpha3: 'bl3' }).execute() )
      }
      await Promise.all(query promises)
      endTime = Date.now()
      console.log(`Time elapsed Coun try insert: ${endTime - startTime}`)
```

**Comment by mathiasmaerker** [comment]
No that is not the issue , just the last version of a long and ugly day ;) I tried `Promise.all` and single insertion with the native driver and that is still way more performant

**Comment by imnotjames** [comment]
How is the single insertion occurring?  Still with the same single prepared statement?

Perhaps enable query logging and inspect what queries are being generated & executed by typeorm?

**Comment by mathiasmaerker** [comment]
You can check #6519 to see the query that is generated by typeorm. And single insertion with the native driver is similliar to the batch query, something along the line:
```javascript
 conn.exec('INSERT INTO "p4c_country" VALUES(?, ?, ?)'), [1,'b2','bl3'], (err, ret)=>{})
```

**Comment by imnotjames** [comment]
How do you test SAP Hana?

It doesn't seem like there's any docker image I can just grab without some sort of store weirdness with dockerhub.

**Comment by mathiasmaerker** [comment]
I will provide you a public repo with a reproduction of the issue. Only question I have is how I can provide you with credentials, since I don't want to post them publicly?

**Comment by Szbuli** [comment]
The native driver uses batch insert. Typeorm will fire each insert sequentially. 1 insert vs 500 insert.

You can check the pool params here: https://github.com/typeorm/typeorm/blob/master/src/driver/sap/SapConnectionOptions.ts

**Comment by mathiasmaerker** [comment]
Yes but even when I use `execute() and Promise.all()` in typeorm and single `await`in native driver, the native driver is 50% faster on an insertion of 100 entries (10 sec compared to 5 sec native) 
so I have a [repo](https://github.com/mathiasmaerker/typeorm-hana-performance-issue) only issue are a credentials file which I would need to provide you

**Comment by imnotjames** [comment]
There's no way to run SAP Hana without paying for it?

**Comment by imnotjames** [comment]
Looks like there's this https://hub.docker.com/r/saphanaspatial/hana-express-for-qgis-tests ? But it's unclear if that would be good for running against..

**Comment by Szbuli** [comment]
You can use sap hana express edition for free. https://developers.sap.com/tutorials/hxe-ua-install-using-docker.html

**Comment by alumni** [comment]
Might be solved by #10198. Basically the `SapQueryRunner` is preparing the statement a sync version of `Connection.prepare` instead of the async/callback one.

**Comment by alumni** [comment]
@mathiasmaerker is this issue still valid?

**Comment by mathiasmaerker** [comment]
@alumni don't know to be honest, I stoped using typeorm some years ago, feel free to close the issue from my side

**Comment by alumni** [comment]
Okay, will be closing this ticket then.

We also encountered some serious performance issues in executing queries and I think it's the same cause as for this issue (which my colleague fixed in the referenced PR).
